server/websocket/src/server.rs

```diff
@@ -189,30 +116,38 @@ where
                             // Obtain the subgraph ID or name that we resolved the request to
                             let subgraph_id = subgraph_id.lock().unwrap().clone().unwrap();
 
+                            // Check if the subgraph is deployed
+                            match store2.is_deployed(&subgraph_id) {
+                                Err(_) |
+                                Ok(false) => {
+                                    error!(logger, "Failed to establish WS connection, subgraph not deployed";
+                                                "subgraph" => subgraph_id.to_string(),
+                                    );
+                                    return Ok(());
+                                }
+                                Ok(true) => (),
+                            }
+
+                            // Get the subgraph schema
+                            let schema = match store2.schema_of(subgraph_id.clone()) {
+                                Ok(schema) => schema,
+                                Err(e) => {
+                                    error!(logger, "Failed to establish WS connection, could not find schema";
+                                                "subgraph" => subgraph_id.to_string(),
+                                                "error" => e.to_string(),
+                                    );
+                                    return Ok(())
+                                }
+                            };
+
                             // Spawn a GraphQL over WebSocket connection
                             let service = GraphQlConnection::new(
                                 &logger,
-                                subgraphs.clone(),
-                                subgraph_id.clone(),
+                                schema,
                                 ws_stream,
                                 graphql_runner.clone(),
                             );
-
-                            // Setup cancelation.
-                            let guard = CancelGuard::new();
```

Comment:
Can't we still do this, but based on subgraph stopped events?

Reply:
I considered this, but it feels like a big mixup of index nodes and query nodes. An approach I thought might be best, is that the store resolver interrupts the stream of responses with an error when the subgraph is no longer deployed. What do you think, would that be an abuse of the resolver?

Reply:
Yep, let's create an issue for it and move on. The resolver approach might work. We do return a stream there that we can decide to cancel when it makes sense.

Reply:
https://github.com/graphprotocol/graph-node/issues/642